### PR TITLE
Smt docs

### PR DESCRIPTION
**Pull request by xwvvvvwx**
- Adds some initial documentation for the smt backend.
- Inserts an implicit `True` invariant to force postcondition checking for all specs with a constructor.

I'm not super in love with the implicit `True` invariant, I mostly added it as a placeholder until I can break the postcondition checking out into a seperate stage.

rendered docs: https://github.com/ethereum/act/blob/smt-docs/docs/smt.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -139,7 +139,7 @@ The types of proofs that can be performed by the various backends can be underst
 | Bytecode                      | Planned | WIP  |         |         |
 | Poststate => postcondition    |         |      | Planned |         |
 | Source level                  |         |      | Planned |         |
-| Contract level invariant      |         |      | Planned | Planned |
+| Contract level invariant      |         |      | Done    | WIP     |
```
Comment (MrChico) [inline README.md:142]: lets add a link to the docs here

File: docs/smt.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+Although SMT solvers are powerful tools there are some important limitations to be aware of:
+
+- The invariants proven can only be inductive in nature, and can only be expressed in terms of state
+  variables and constructor arguments.
+- SMT solvers do not support the exponentiation operation.
+- Bytestrings are encoded using the theory of sequences. This is a non standard smtlib extension and
+  is currently only supported by `z3`. Behaviours that use the `string`, `bytes`, or `bytesX`
+  (e.g. `bytes32`) types will not be able to make use of the `cvc4` backend.
+- The solver may be unable to prove properties about behaviours making extensive use of non linear
+  arithmetic (i.e. multiplication or division by a symbolic value). This is an inherant limitation
```
Comment (MrChico) [inline docs/smt.md:31]:
```suggestion
  arithmetic (i.e. multiplication or division by a symbolic value). This is an inherent limitation
```

File: docs/smt.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
+# SMT Backend
+
+The SMT backend for act is currently able to automatically prove that:
```
Comment (leonardoalt) [inline docs/smt.md:3]:
```suggestion
The SMT backend for Act is currently able to automatically prove that:
```

File: docs/smt.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+- post conditions are implied by pre conditions
```
Comment (leonardoalt) [inline docs/smt.md:5]: is this working atm?

File: docs/smt.md [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+To perform these proofs, you can simply run `act prove --file <PATH_TO_SPEC>` against your spec.
+
+`act prove` also accepts the following configuration flags:
+
+- `--solver`: you can choose to use `cvc4` or `z3` as the solver backend. The default is `z3`.
+  Sometimes `cvc4` may be able to prove things that `z3` cannot (and vice versa). You can also
+  prove the same properties with multiple solvers to gain confidence that the proofs are not
+  affected by a bug in the solver itself.
+- `--smttimeout`: the timeout given for each smt query. This is set to 20s by default.
+- `--debug`: this prints the raw query dispatched to the smt solver to stdout. Note that these
```
Comment (leonardoalt) [inline docs/smt.md:17]:
```suggestion
- `--debug`: this prints the raw query dispatched to the SMT solver to stdout. Note that these
```

File: docs/smt.md [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+   queries are machine generated by the `sbv` library and so are fairly challening to inspect manually.
```
Comment (leonardoalt) [inline docs/smt.md:18]:
```suggestion
   queries are machine generated by the `sbv` library and so are fairly challenging to inspect manually.
```

File: docs/smt.md [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+
+## Examples
+
+### Automatic Market Maker
+
+As an example consider the following specification of a simplified uniswap style automatic market
+maker. For the purposes of this example we focus only on the core logic, and ignore e.g.
+interactions with the underlying tokens.
+
+The `Amm` has two state variables representing it's underlying token blances, `reserve0` and
```
Comment (leonardoalt) [inline docs/smt.md:42]:
```suggestion
The `Amm` has two state variables representing its underlying token balances, `reserve0` and
```

File: docs/smt.md [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+
+    uint256 reserve0 := 1000
+    uint256 reserve1 := 1000
+
+invariants
+
+    1000 * 1000 <= reserve0 * reserve1
+```
+
+Next, we define the two `swap` methods, that allow a user to exchange tokens, these methods take an
```
Comment (leonardoalt) [inline docs/smt.md:63]:
```suggestion
Next, we define the two `swap` methods, that allow a user to exchange tokens. These methods take an
```

File: docs/smt.md [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+storage
+
+    reserve0 => (reserve0 * reserve1) / (reserve1 + amt) + 1
+    reserve1 => reserve1 + amt
+```
+
+### Underpowered Invariant
+
+Due to the inductive nature of the proof, there are some true invariants that the SMT backend is
+unable to prove. For example, consider the following state machine:
```
Comment (leonardoalt) [inline docs/smt.md:140]: I'm not sure about this part. It's not like the SMT solver wouldn't be able to prove that `x < 9` is an invariant, the backend is just not interested in non-inductive invariants.

Comment (leonardoalt) [reply docs/smt.md:140]: To clarify: I think the example is good to distinguish inductive vs non-inductive invariants, I'm just not sure about the wording above.

File: docs/smt.md [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+
+If all of the queries for an invariant claim return `unsat`, then two properties have been proved
+about the transition system:
+
+1. The invariant holds over the post state
+2. The postconditions hold for every method level behaviour
+
+## Future Extensions
+
+In the future we hope to leverage smt solvers to prove the following claims:
```
Comment (leonardoalt) [inline docs/smt.md:220]:
```suggestion
In the future we hope to leverage SMT solvers to prove the following claims:
```

File: docs/smt.md [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+In the case above, the invariant states that `x < 9`, and if this is assumed as a precondition, then
+the `x == 7` branch in `j()` is still reachable.
+
+We can fix this by strengthening the invariant to make the `x == 7` branch unreachable, in the case
+of the spec above, an invariant of `x < 7` is sufficient, although `x < 3` is the strongest
+invariant of this form that holds over the full transition system.
+
+## Implementation Strategy
+
+For each invariant claim, act builds an individual SMT query for each behaviour in the transtion
```
Comment (leonardoalt) [inline docs/smt.md:194]:
```suggestion
For each invariant claim, Act builds an individual SMT query for each behaviour in the transtion
```
or \`act\` everywhere?

**Review by leonardoalt — COMMENTED** [review]
Nice!

File: docs/smt.md (hunk 4 above)
Comment (xwvvvvwx) [reply docs/smt.md:5]: It's a little messy, but if there is a construcor then yes. @MrChico modified the invariant checking to also check postconditions, and in this PR I insert an implicit `True` invariant to all specs with a constructor.

I would like to pull the postcondition checking out into it's own pass so we can simplify the queries and present nicer error messages to users.

File: docs/smt.md (hunk 9 above)
Comment (xwvvvvwx) [reply docs/smt.md:140]: Any suggestions for better phrasing? I already have "due to the inductive nature of the proof" in the sentance above?

**Comment by MrChico** [comment]
I think the example can live in its own header, not under the `limitations` header, otherwise this lookks great!

**Comment by xwvvvvwx** [comment]
The examples are their own header? Or do you mean I should move the limitations underneath the examples?

**Comment by MrChico** [comment]
Both

**Pull request merged**